### PR TITLE
Fix unreachable code detection in match statements

### LINKED ISSUE #30458: Faulty "Unreachable code" warning on match statement containing a return
**Godot version:**
v3.1.1.stable.official

**OS/device including version:**
Ubuntu 18.04.2 LTS

**Issue description:**
The editor throws a "Unreachable code" warning after a "match" statement that contains a single unused return. 

_I am aware that this behavior is probably not really new, but the other issues on that are closed while this type of warning still occurs._ (see #25955, #26097, #28360)

**Steps to reproduce:**
Run the following script:

```
extends Node2D

func _ready():
	var selector = 0
	match selector:
		1:
			return
	
	print("we reached the unreachable")
```

The output says:
```
** Debug Process Started **
OpenGL ES 3.0 Renderer: GeForce GTS 450/PCIe/SSE2
we reached the unreachable
```
But the editor still receives a warning:
```
W 0:00:00:0765   Unreachable code (statement after return) in function '_ready()'.
  <Source>       test.gd:9
```

### LINKED ISSUE #30490: No unreachable code warning with match block when Wildcard Pattern is used earlier than last case
**Godot version:**

Latest master https://github.com/godotengine/godot/commit/9496e29326ec483a77b1246874623f09d74b9363

**OS/device including version:**

Windows 10 1903 64 bit.

**Issue description:**

There are no unreachable code warning with match block when Wildcard Pattern is anytime before last case.
```
        var selector = "1"# or anything else
	match selector:
                _:
                       print("Super-case")
		"1":
			print("Mr. Unreachable")
		2:
			print("Mrs. Unreachable")
		_:
			print("Jr. Unreachable(bad-super-case-knockoff?)")
```
This always prints "Super-case". So, the lines after it should be unreachable.
Though I hardly think anyone would use wild card match pattern before the end but some beginners just might.

**Steps to reproduce:**
1. Just run the above code for different values of selector.
2. Only the super-case is printed no matter what the value is given to selector.
(Cry in agony as the unreachable family is never reached).

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by ayuusweetfish**
Fixes #30458 and fixes #30490.

### CODE REVIEW

**Comment by kawa-yoiko** [comment]
Force updated due to build failing with `DEBUG_ENABLED` turned off.

**Comment by kawa-yoiko** [comment]
Force pushed due to clang-format reports.

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -2259,6 +2269,8 @@ void GDScriptParser::_parse_pattern_block(BlockNode *p_block, Vector<PatternBran
 			catch_all = catch_all || pt == PatternNode::PT_WILDCARD;
 		}
 
+		catch_all_appeared |= catch_all;
```
Comment (bojidar-bg) [inline modules/gdscript/gdscript_parser.cpp]: (non-blocking) For consistency, this should probably be rewritten as the statement 3 lines above:
```cpp
catch_all_appeared = catch_all_appeared || catch_all;
```

Comment (kawa-yoiko) [reply modules/gdscript/gdscript_parser.cpp]: Oh sure, updated.

**Review by bojidar-bg — APPROVED** [review]
Looks good, I guess

**Comment by akien-mga** [comment]
Thanks!

**Comment by Anutrix** [comment]
This fixes #30490 but still slightly wrong error message. It says
`Line X (UNREACHABLE_CODE):Unreachable code (statement after return) in function '_some_func()'`
even when there is no return. I think the error should be
`Line X (UNREACHABLE_CODE):Unreachable code (statement after wildcard case) in function '_some_func()'`

**Comment by kawa-yoiko** [comment]
Ah indeed, didn't notice > <

We can either remove that "statement after return" explanation, or separate the two cases to be e.g. `UNREACHABLE_AFTER_RETURN` and `UNREACHABLE_AFTER_WILDCARD`. Any ideas?

**Comment by Anutrix** [comment]
Separating the two would be better in my opinion as it often isn't apparent that wildcard might cause unreachable code.

**Pull request merged**